src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,330 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { QueuedSourceBuffer } from "../../../core/source_buffers";
+import Adaptation from "../../../manifest/adaptation";
+import arrayFind from "../../../utils/array_find";
+import arrayFindIndex from "../../../utils/array_find_index";
+import PPromise from "../../../utils/promise";
+import appendInitSegment from "./append_init_segment";
+import loadArrayBufferData from "./load_array_buffer_data";
+import prepareSourceBuffer from "./media_source";
+
+interface IThumbnailInfo {
+  start: number;
+  duration: number;
+  mediaURL: string;
+}
+
+interface IThumbnailTrack {
+  thumbnailInfos: IThumbnailInfo[];
+  initURL: string;
+  codec: string;
+}
+
+const MAXIMUM_MEDIA_BUFFERED = 2;
+
+/**
+ * This tool, as a supplement to the RxPlayer, intent to help creating thumbnails
+ * from a video source.
+ *
+ * From a given video adaptation, the tools will extract a "thumbnail track",
+ * either from a trickmode track (whose light chunks are adapted from such use case)
+ * or direclty from the media content.
+ */
+export default class VideoThumbnailLoader {
+  private readonly _thumbnailVideoElement : HTMLVideoElement;
+  private readonly _onlyTrickMode : boolean;
+  private readonly _videoSourceInfos: Promise<{
+    videoSourceBuffer: QueuedSourceBuffer<ArrayBuffer>;
+    disposeMediaSource: () => void;
+  }>;
+  private readonly _bufferedDataCache: Array<{
+    start: number;
+    end: number;
+    data: ArrayBuffer;
+  }>;
+
+  private  _initSegment: ArrayBuffer|null;
+  private _thumbnailTrack : IThumbnailTrack;
+  private _currentCallback: Promise<unknown>|null;
+  private _nextCallback: null|(() => Promise<unknown>);
+
+  constructor(
+    videoElement: HTMLVideoElement,
+    adaptation: Adaptation,
+    onlyTrickmode: boolean = true
```

Comment:
No need for onlyTrickMode. Just let the UI give the right Adaptation. Either the trickmode one or a video one

Reply:
We could also just only give the corresponding `Representation`, but I'm not sure yet.